### DISCUSSION TITLE
How do you calculate GPA in C++?

### DISCUSSION THREAD

**Discussion by ghost** [community/community#54048] [Programming Help] [answered]
### Body

I made a program to calculate your mark awhile back and now I’m adding the gpa part but I don’t know how to do it.
So, how do I calculate GPA in C++?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by ghost** [comment]
Here’s my code https://github.com/PyTerminal/MarkCalculator/blob/main/main.cpp

**Comment by FEsoftdev** [comment, marked as answer]
To calculate GPA in C++, you will need to first determine the grade point value for each letter grade, then calculate the GPA by taking the weighted average of the grade point values.

```
#include <iostream>
#include <string>

using namespace std;

// function to get the grade point value for a letter grade
double getGradePoint(char letterGrade) {
    switch (letterGrade) {
        case 'A':
            return 4.0;
        case 'B':
            return 3.0;
        case 'C':
            return 2.0;
        case 'D':
            return 1.0;
        default:
            return 0.0;
    }
}

int main() {
    int numCourses;
    double totalCredits = 0.0;
    double totalGradePoints = 0.0;

    cout << "Enter the number of courses: ";
    cin >> numCourses;

    for (int i = 1; i <= numCourses; i++) {
        double credits;
        char letterGrade;

        cout << "Enter the credits for course " << i << ": ";
        cin >> credits;

        cout << "Enter the letter grade for course " << i << ": ";
        cin >> letterGrade;

        // calculate the grade point value for the letter grade
        double gradePoint = getGradePoint(letterGrade);

        // calculate the grade points for the course
        double courseGradePoints = gradePoint * credits;

        // add the course grade points and credits to the total
        totalGradePoints += courseGradePoints;
        totalCredits += credits;
    }

    // calculate the GPA
    double gpa = totalGradePoints / totalCredits;

    cout << "Your GPA is: " << gpa << endl;

    return 0;
}
```

This program prompts the user to enter the number of courses they have taken, and for each course, it prompts for the credits and letter grade.

**Comment by neoalabs** [comment]
You’ll basically want to map each grade to its point value (e.g., A = 4.0, B = 3.0, etc.), multiply by credits if applicable, and then take the average.

If you're testing your implementation, it can help to compare your output with something external—I’ve used CalculatorWallah before just to sanity-check GPA calculations while debugging.

Here is code - 
def calculate_gpa(grades, credits):
    # Grade → Points mapping
    grade_points = {
        "A+": 4.0, "A": 4.0, "A-": 3.7,
        "B+": 3.3, "B": 3.0, "B-": 2.7,
        "C+": 2.3, "C": 2.0, "C-": 1.7,
        "D+": 1.3, "D": 1.0, "F": 0.0
    }

    total_points = 0
    total_credits = 0

    for grade, credit in zip(grades, credits):
        grade = grade.upper()
        if grade in grade_points:
            total_points += grade_points[grade] * credit
            total_credits += credit
        else:
            print(f"Invalid grade: {grade}")

    if total_credits == 0:
        return 0

    return round(total_points / total_credits, 2)


# Example usage
grades = ["A", "B+", "A-", "C"]
credits = [3, 4, 3, 2]

gpa = calculate_gpa(grades, credits)
print("GPA:", gpa)
